### ISSUE TITLE
vmtoolsd crash because non allocated pointer passed to realloc in services/plugins/guestInfo/diskInfo.c

### ISSUE THREAD

**Issue by nkshirsagar** [vmware/open-vm-tools#257]
rhel bz filed at https://bugzilla.redhat.com/show_bug.cgi?id=1583524

(gdb) bt
#0  0x00007f358390c1f7 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007f358390d8e8 in __GI_abort () at abort.c:90
#2  0x00007f358394bf47 in __libc_message (do_abort=do_abort@entry=2, 
    fmt=fmt@entry=0x7f3583a58608 "*** Error in `%s': %s: 0x%s ***\n")
    at ../sysdeps/unix/sysv/linux/libc_fatal.c:196
#3  0x00007f3583951b54 in malloc_printerr (action=<optimized out>, 
    str=0x7f3583a587d0 "malloc(): smallbin double linked list corrupted", 
    ptr=<optimized out>, ar_ptr=<optimized out>) at malloc.c:5023
#4  0x00007f3583954df7 in _int_malloc (
    av=av@entry=0x7f3583c93760 <main_arena>, bytes=bytes@entry=225)
    at malloc.c:3385
#5  0x00007f358395587c in _int_realloc (
    av=av@entry=0x7f3583c93760 <main_arena>, oldp=oldp@entry=0x55a14f23aa00, 
    oldsize=oldsize@entry=128, nb=nb@entry=240) at malloc.c:4290
#6  0x00007f3583957752 in __GI___libc_realloc (oldmem=0x55a14f23aa10, 
    bytes=bytes@entry=232) at malloc.c:3028
#7  0x00007f3587d20d39 in UtilSafeRealloc0 (ptr=<optimized out>, <-----
    size=size@entry=232) at utilMem.c:128
#8  0x00007f357a45e077 in GuestInfoGetDiskInfoWiper () at diskInfo.c:121
#9  0x00007f357a45e175 in GuestInfo_GetDiskInfo () at diskInfoPosix.c:45
#10 0x00007f357a45c975 in GuestInfoGather (
    data=data@entry=0x55a14dbc2100 <gState+96>) at guestInfoServer.c:381

I think the root cause is we have allocated di but not di->partitionlist

di = Util_SafeCalloc(1, sizeof *di); <--- 

Then we pass to reallocate this 

newPartitionList = Util_SafeRealloc(di->partitionList,	                                             (partCount + 1) *	                                           sizeof *di->partitionList);

Note that *di->partitionList is not malloc ed

So we are passing a non mallocd pointer to reallocate.

A possible patch would be to change

-------------------------------------------------------------------------

   /* Get partition list. */
   if (!WiperPartition_Open(&pl, FALSE)) {
      g_warning("GetDiskInfo: ERROR: could not get partition list\n");
      return FALSE;
   }

   di = Util_SafeCalloc(1, sizeof *di);
   partNameSize = sizeof (di->partitionList)[0].name;

------------------------------------------------------------------------

to 
------------------------------------------------------------------------
   /* Get partition list. */
   if (!WiperPartition_Open(&pl, FALSE)) {
      g_warning("GetDiskInfo: ERROR: could not get partition list\n");
      return FALSE;
   }

   di = Util_SafeCalloc(1, sizeof *di);                      <----- patch
   di->partitionList = Util_SafeCalloc(1,sizeof *di->partitionList)

   partNameSize = sizeof (di->partitionList)[0].name;

------------------------------------------------------------------------

So we may fix this crash by calloc'ing it and preventing the garbage pointer  being passed in to realloc()

Even though it seems that it has reallocated successfully most of the time, hiding the bug, note the undefined behavior mentioned in realloc's documentation here though, as pasted earlier too)

----------------------------------------------------------------------------

"
realloc deallocates the old object pointed to by ptr and returns a pointer to a new object that has the size specified by size. The contents of the new object is identical to that of the old object prior to deallocation, up to the lesser of the new and old sizes. Any bytes in the new object beyond the size of the old object have indeterminate values.

The point to note is that realloc() should only be used for dynamically allocated memory. If the memory is not dynamically allocated, then behavior is undefined. <-------

----------------------------------------------------------------------------

We just use partNameSize to get the size of the variable so thats fine and doesnt crash later, (note sizeof is an operator, and not a function so its basically taking the size of the static variable "name" in the structure, which is 100 bytes array.

(gdb) ptype di
type = struct _DiskInfo {
    unsigned int numEntries;
    PPartitionEntry partitionList;
} *

(gdb) ptype PPartitionEntry
type = struct _PartitionEntry {
    uint64 freeBytes;
    uint64 totalBytes;
    char name[100]; <---- so we get 100 here since this is statically allocated
} *

   partNameSize = sizeof (di->partitionList)[0].name; <-- so this works fine even though di->partitionList is technically garbage at this point.

(gdb) p partNameSize
$3 = 100  <---
(gdb) 

Later, we do

         newPartitionList = Util_SafeRealloc(di->partitionList,
                                             (partCount + 1) *
                                             sizeof *di->partitionList);

         if (strlen(part->mountPoint) + 1 > partNameSize) {
            g_warning("GetDiskInfo: ERROR: Partition name buffer too small\n");
            goto out;
         }

Then everything else will work fine, if the realloc actually worked OK, since now we are dealing with a newly allocated pointer for di->partitionList. We then index into it using partEntry, and then strcpy() stuff into it. Thats OK.

         newPartitionList = Util_SafeRealloc(di->partitionList,
                                             (partCount + 1) *
                                             sizeof *di->partitionList);

         partEntry = &newPartitionList[partCount++];
         Str_Strcpy(partEntry->name, part->mountPoint, partNameSize); 

However, since the behavior is undefined, its better to allocate the inner pointer before passing it to realloc, since realloc() expects a malloc'd pointer.

Basically if you have, as an example,

struct x
{
  char *innerpointer;
}*pointer_to_x;

to allocate this structure correctly, we must do pointer_to_x=malloc(...); and then we must also do pointer_to_x->innerpointer=malloc(...) without which we have a dangling pointer in the structure, which in this case we are passing to realloc()

Thanks,
nikhil.

**Comment by nkshirsagar** [comment]
Here is a patch which I think will fix this vmtoolsd crash,

[root@nkshirsa open-vm-tools]# git diff services/plugins/guestInfo/diskInfo.c
diff --git a/open-vm-tools/services/plugins/guestInfo/diskInfo.c b/open-vm-tools/services/plugins/guestInfo/diskInfo.c
index d178944..537326a 100644
--- a/open-vm-tools/services/plugins/guestInfo/diskInfo.c
+++ b/open-vm-tools/services/plugins/guestInfo/diskInfo.c
@@ -96,6 +96,8 @@ GuestInfoGetDiskInfoWiper(Bool includeReserved)  // IN
    }
 
    di = Util_SafeCalloc(1, sizeof *di);
+   /* bz 1583524:  must allocate the inner pointer before passing to reallocate */
+   di->partitionList = Util_SafeCalloc(1,sizeof *di->partitionList); 
    partNameSize = sizeof (di->partitionList)[0].name;
 
    DblLnkLst_ForEach(curr, &pl.link) {

**Comment by ravindravmw** [comment]
As I explained in the Red Hat bug, I don't think this fix makes any difference because calloc() would initialize di->partitionList to NULL. We need to debug it more because it seems like memory corruption.

**Comment by nkshirsagar** [comment]
Hello,

I will respond to your comment on the bz in a bit.

Regards,
Nikhil.

**Comment by nkshirsagar** [comment]
Hello,

As explained in  https://bugzilla.redhat.com/show_bug.cgi?id=1583524#c29 we cannot fully trust calloc to null out inner pointer properly.

So it's not just the unallocated pointer being passed to reallocate. It may also be that the pointer in this case was not noticed to be null.. and freed. 

I can try to reproduce this issue in a test application and try to get a core dump in a similar kind of memory allocation. Shall keep you posted.

Nikhil.
